services/queue/src/main/java/tech/pegasys/pantheon/services/queue/InMemoryBigQueue.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.services.queue;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class InMemoryBigQueue<T> implements BigQueue<T> {
+  private final AtomicLong size = new AtomicLong(0);
+  private final Queue<T> internalQueue = new ConcurrentLinkedQueue<>();
```

Comment:
Removed the separate size tracking.  I _think_ that the way `isEmpty` is being used in the downloader should be okay because the flow is:  
- send request and increment outstandingRequest count
- receive response and process it
  - queue additional requests
- decrement outstandingRequests and check if the queue is empty

Assuming that once requests are queued (enqueue returns), an updated count should be visible to other threads, I think that flow works.  Does that seem right to you?

Reply:
As long as the outstanding request count is incremented *before* the request is actually sent (which it was from memory) then yes I agree that works.